File: sampleCode/CSharp/ConsoleApp/Workflows/CreateProjectWorkflow.cs
namespace ConsoleApp.Workflows;

public class CreateProjectWorkflow : IWorkflow
{
    /// <summary>
    /// This Example Workflow follows along with `api\impact\workflows\CreateProject.md`
    /// </summary>
    public static void Examples()
    {
        /* Projects are the top level of organization for an analysis and contain the specifics of the Aggregation Scheme, Household Set, MRIO status, and folder placement.
         * This example shows off the most basic of Project Creation workflows
         * Please see the main Impact Readme at `api\impact\readme.md` and `support.implan.com` for more information.
         */
        
        
        /* The first step to create a Project is to find the Aggregation Scheme */

        // Get a list of all valid Aggregation Schemes
        AggregationScheme[] aggregationSchemes = AggregationSchemes.GetAggregationSchemes();

        // Choose the one that you want to use
        AggregationScheme implan546AggScheme = aggregationSchemes.First(agg => agg.Description == "546 Unaggregated");
        // You will need its Aggregation Scheme Id
        int aggregationSchemeId = implan546AggScheme.Id;    // 8
        // As well as a valid Household Set Id
        int householdSetId = implan546AggScheme.HouseholdSetIds.First();    // 1

        
        /* Once you have chosen your Aggregation Scheme Id and one of its Household Set Ids, you can create your Project */
        
        // Define the project with its required properties
        Project project = new Project
        {
            // The title must be unique, so we're using the timestamp here
            Title = $"ProjectWorkflow - {DateTime.Now:s}",
            AggregationSchemeId = aggregationSchemeId,
            HouseholdSetId = householdSetId,
        };
        
        // This endpoint creates a Project and returns basic information about it, including a unique identifier `projectId` (guid) to be used in other API requests.
        // Note: Calling this endpoint returns a fully-hydrated Project, so we re-assign it here
        project = Projects.Create(project);
        
      
        // With a Project created, you will next need to determine an industry code to use for your event.
        // A list of industries that can be utilized for your analysis.
        // Events created later will require reference to one of the industry codes (`code`) returned here.
        
        // Industries are seperated into different Industry Sets
        IndustrySet[] industrySets = IndustrySets.GetIndustrySets();
        // Choose the one that describes your industry
        IndustrySet implan546IndustriesSet = industrySets.First(s => s.Description == "546 Industries");     // 8
        
        // You need to get an Industry Code for the Impact Event -- which can be further filtered by an Industry Set
        IndustryCode[] industryCodes = IndustryCodes.GetIndustryCodes(aggregationSchemeId, 
            industrySetId: implan546IndustriesSet.Id);
        // Choose the one that best fits
        IndustryCode industryCode = industryCodes.First(c => c.Description == "Oilseed farming");   // 1
        
        
        /* There are many types of Industry Event, each requires a different Payload to be sent to the AddEvent endpoint */
        
        // This endpoint shows all valid Event Types for a given Project
        string[] eventTypes = Events.GetEventsTypes(project.Id);
        // See the main Impact Readme or support for more details on different event types
        
        
        // An Industry Output Event
        IndustryOutputEvent industryOutputEvent = new IndustryOutputEvent()
        {
            Title = "Industry Output Event",
            IndustryCode = industryCode.Code,
            Output = 100_000.00,
        };
        
        // An Industry Impact Analysis
        IndustryImpactAnalysisEvent industryImpactAnalysisEvent = new IndustryImpactAnalysisEvent()
        {
            Title = "Industry Impact Analysis Event",
            IndustryCode = industryCode.Code,
            IntermediateInputs = 500_000,
            EmployeeCompensation = 250_000,
            ProprietorIncome = 50_000,
            WageAndSalaryEmployment = 4,
            ProprietorEmployment = 1,
            TotalEmployment = 5,
            TotalLaborIncome = 300_000,
            OtherPropertyIncome = 100_000,
            TaxOnProductionAndImports = 100_000,
            SpendingPatternDatasetId = 87,
            SpendingPatternValueType = SpendingPatternValueType.IntermediateExpenditure,
        };

        // Add the events to the Project we just created -- will return a new Event with information filled in
        industryOutputEvent = Events.AddEvent(project.Id, industryOutputEvent);
        industryImpactAnalysisEvent = Events.AddEvent(project.Id, industryImpactAnalysisEvent);
        
        
        /* Now that Event(s) have been added, it is time to find the Region(s) that are to be used in the Impact */
        // See Regions.cs, CombinedRegionWorkflow.cs, and RegionalWorkflow.cs for more examples on searching regions
        
        // Regions must be associated with a Data Set
        // Get a list of all valid Data Sets for the Aggregation Scheme
        DataSet[] datasets = DataSets.GetDataSets(aggregationSchemeId);
        // Choose the one you would like to use -- must be compatible with your chosen HouseholdSetId!
        DataSet dataset = datasets.First(d => d.Description == "2022");
        
        // For this example, we're going to search through the Child Regions of the US for a particular state
        Region[] stateRegions = Regions.GetRegionChildren(aggregationSchemeId, dataset.Id, regionType: "State");
        Region oregonStateRegion = stateRegions.First(s => s.Description == "Oregon");
        // We need the HashId specifically
        string oregonStateHashId = oregonStateRegion.HashId;    // 15b869ZOxy
        
        
        /* Adding Groups
         * Groups represent the Region and timeframe in which an Event takes place.
         * Use the following endpoints to arrange Groups in your project.
         * Reference the `ProjectId` and `EventId`s from the created Project and Events where required.
         */
        
        // Define the group
        Group group = new Group()
        {
            ProjectId = project.Id,
            Title = "Sample Group",
            DatasetId = dataset.Id,
            DollarYear = 2024,
            // Must specify at least one regional identifier
            HashId = oregonStateHashId,
            // Must add at least one Event
            GroupEvents = [
                new GroupEvent(eventId: industryOutputEvent.Id),
                new GroupEvent(eventId: industryImpactAnalysisEvent.Id),
            ],
        };

        // Note: Calling this endpoint returns a fully-hydrated Group, so we re-assign it here
        group = Groups.AddGroupToProject(project.Id, group);
        
        
        // Now that the Project has been fully defined, see
        RunImpactAnalysisWorkflow.ProjectId = project.Id;
        RunImpactAnalysisWorkflow.Examples();
        // for ways to run an Analysis and view the Results
    }
}

File: sampleCode/R/CreateProjectWorkflow.R
source(auth_env$Authentication)
source(auth_env$AggregationSchemeEndpoints)
source(auth_env$Rest)
source(auth_env$ProjectEndpoints)
source(auth_env$IndustrySetEndpoints)
source(auth_env$IndustryCodeEndpoints)
source(auth_env$EventEndpoints)
source(auth_env$DataSetEndpoints)
source(auth_env$RegionEndpoints)
source(auth_env$GroupEndpoints)
source(auth_env$Iworkflow)
source(auth_env$RunImpactAnalysisWorkflow)



# regions_instances <- new("Regions")
# 
# regions <- GetRegionTypes(regions_instances)
# 
# print(regions)

setClass("CreateProjectWorkflow",
         contains = "Iworkflow")

# Create a generic function for Create
#setGeneric("Examples", function(object) standardGeneric("Examples"))

# Define a method for the Create function with the CreateProjectWorkflow class
setMethod("Examples", signature = "CreateProjectWorkflow", function(object){
  
  AllSchemes <- new("AggregationSchemeEndpoints")
  aggregationSchemes <- GetAggregationSchemes(AllSchemes,IndustrySetId=8)
  
  print(aggregationSchemes)
  
  
  implan546AggScheme <- aggregationSchemes[aggregationSchemes$description=="546 Unaggregated",]
  
  aggregationSchemeId <- implan546AggScheme$id
  
  householdSetId <- as.integer(implan546AggScheme$householdSetIds)
  
  Title = paste0("ProjectWorkflow ",Sys.time())
  
  # Create an instance of `project_instance`
  project_instance <- new("ProjectEndpoints",
                          Title = paste0("ProjectWorkflow ",Sys.time()),
                          AggregationSchemeId = aggregationSchemeId,
                          HouseholdSetId = householdSetId[1]
  )
  
  print(project_instance)
  
  # Call the Create method
  project <- Create(project_instance)
  
  project
  
  industries <- new("IndustrySetEndpoints")
  
  industrySets <- GetIndustrySets(industries)
  
  industrySets
  
  implan546IndustriesSet <- industrySets[industrySets$description=="546 Industries",][1,] #[1,] accessing first row
  
  industryCodes <- new("IndustryCodeEndpoints")
  
  IndustryCodes <- GetIndustryCodes(industryCodes,aggregationSchemeId,implan546IndustriesSet$id)
  
  IndustryCodes
  
  IndustryCode <- IndustryCodes[IndustryCodes$description=="Oilseed farming",][1,] #[1,] accessing first row
  
  IndustryCode
  
  Event_class <- new("EventEndpoints")
  
  eventTypes <- GetEventsTypes(Event_class,project$id)
  
  eventTypes
  
  # Create an instance of `IndustryOutputEvent`
  industry_output_event <- new("IndustryOutputEvent",
                               Output = 100000,
                               IndustryCode = IndustryCode$code,
                               Title = "Industry Output Event"
  )
  
  print(industry_output_event)
  
  
  industryOutputEvent <- AddEvent(industry_output_event,
                                  projectGuid = project$id)
  
  industryOutputEvent
  
  
  industryImpactAnalysisEvent <- new("IndustryImpactAnalysisEvent",
                                     Title = "Industry Impact Analysis Event",
                                     IndustryCode = IndustryCode$code,
                                     IntermediateInputs = 500000,
                                     EmployeeCompensation = 250000,
                                     ProprietorIncome = 50000,
                                     WageAndSalaryEmployment = 4,
                                     ProprietorEmployment = 1,
                                     TotalEmployment = 5,
                                     TotalLaborIncome = 300000,
                                     OtherPropertyIncome = 100000,
                                     TaxOnProductionAndImports = 100000,
                                     SpendingPatternDatasetId = as.integer(87),
                                     SpendingPatternValueType = as.character(SpendingPatternValueType[SpendingPatternValueType == "IntermediateExpenditure"])
  )
  
  industryImpactAnalysisEvent <- AddEvent(industryImpactAnalysisEvent,
                                           projectGuid = project$id)
  
  industryImpactAnalysisEvent
  
  datasets_instance <- new("DataSetEndpoints")
  
  datasets <- GetDataSets(datasets_instance,aggregationSchemeId)
  
  datasets
  
  dataset <- datasets[datasets$description=="2022",][1,] #[1,] accessing first row
  
  dataset
  
  Regions_instance <- new("RegionEndpoints")
  
  stateRegions <- GetRegionChildren(Regions_instance, aggregationSchemeId = aggregationSchemeId ,
                                    dataSetId = dataset$id ,regionType = "State")
  stateRegions
  
  oregonStateRegion <- stateRegions[stateRegions$description == "Oregon",][1,] #[1,] accessing first row
  
  oregonStateHashId <- oregonStateRegion$hashId
  
  oregonStateHashId
  
  print(paste0(project$id," ", dataset$id," ", oregonStateHashId," ", industryOutputEvent$id," ", industryImpactAnalysisEvent$id))
  
  # Create an instance of `GroupEndpoints`
  group_instance <- new("GroupEndpoints",
                        ProjectId = project$id,
                        Title = "Sample Group",
                        DatasetId = dataset$id,
                        DollarYear = as.integer(2024),
                        # Must specify at least one regional identifier
                        HashId = oregonStateHashId,
                        
                        # Must add at least one Event
                        GroupEvents = list(GroupEvent(new("Event", eventId = industryOutputEvent$id)), 
                                           GroupEvent(new("Event", eventId = industryImpactAnalysisEvent$id)))
  )
  
  print(group_instance)
  
  
  group <- AddGroupToProject(projectGuid = project$id, group_instance)
  
  group
  
  RunImpactAnalysisWorkflow_instance <- new("RunImpactAnalysisWorkflow")
  
  print(Examples(RunImpactAnalysisWorkflow_instance,ProjectId = project$id))
})

